File: csharp2ts/Creature.cs
using System;
using System.Drawing;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MicroTerraformer2
{


    /// <summary>
    /// manages creature properties from colors
    /// </summary>
    class Creature
    {
        //
        //  DEFINES
        //
        

        // trophicLevel
        public const int SOIL = 0;
        public const int PLANT = 1;
        public const int HERBIVORE = 2;
        public const int CARNIVORE = 3;

        // attackResult
        public const int DRAW = 0;
        public const int KILL = 1;
        public const int DIE = 2;

        public static Color deadCreatureColor = Color.FromArgb(255, 255, 255, 128);    // equivalent to less defense soil
        

        //
        //  SIMULATION PARAMETERS
        //

        // consumes this energy every cycle
        const float ENERGY_CONSUMPTION_PLANTS = 1.0f;
        const float ENERGY_CONSUMPTION_HERBIVORES = 2.0f;       // 0.5
        const float ENERGY_CONSUMPTION_CARNIVORES = 2.0f;

        const float ENERGY_INCREMENT_KILLING = 2f;

        public const int DEFENSE_LEVEL_MAX = 150;
        public const int ATTACK_LEVEL_MAX = 150;
        public const int ENERGY_LEVEL_MAX = 100;
        public const int ENERGY_LEVEL_COLOR_MIN = 155;              // energyLevel = color - ENERGY_LEVEL_COLOR_MIN
        public const int ENERGY_LEVEL_COLOR_NOT_ACTIVATED = 154;   // 154 (energyLevel = -1) means "not activated"


        public const int ENERGY_LEVEL_AT_BIRTH = 50;

        public const float PLANT_REPRODUCTION_PROBABILITY = 0.20f;        // recommended 0.1 - 0.5
        public const float HERBIVORE_REPRODUCTION_PROBABILITY = 0.30f;    // recommended 0.15 - 
        public const float CARNIVORE_REPRODUCTION_PROBABILITY = 0.30f;

        // probability to move after killing a creature
        public const float PLANT_MOVE_PROBABILITY = 0f;        
        public const float HERBIVORE_MOVE_PROBABILITY = 0.20f;     
        public const float CARNIVORE_MOVE_PROBABILITY = 0.40f;


        public struct lifeParameters
        {
            public int trophicLevel;
            public int attackLevel;
            public int defenseLevel;
            public float energyLevel;
            public Color myColor;
        }

        public lifeParameters me;

        private Random rand = new Random();

        // Consumes energy and returns color for new state
        // If herbivores or carnivores are not still activated, don't consume energy

        public void BeginIteration(Color myColor)
        {

            me = LifeParametersFromColor(myColor);

            // not "living creatures" doesn't have metabolism...
            if (me.trophicLevel == SOIL)
                return;
            if (this.IsActivated()==false)
                return;

            // consumes energy
            switch (me.trophicLevel)
            {
                case PLANT:
                    me.energyLevel -= ENERGY_CONSUMPTION_PLANTS;
                    break;
                case HERBIVORE:
                    me.energyLevel -= ENERGY_CONSUMPTION_HERBIVORES;
                    break;
                case CARNIVORE:
                    me.energyLevel -= ENERGY_CONSUMPTION_CARNIVORES;
                    break;
            }

            me.energyLevel = me.energyLevel < 0 ? 0 : me.energyLevel;

            // update my color with new energy level
            me.myColor = ColorFromLifeParameters(me);
        }

        // returns attackResult

        public int Attack(Color itsColor)
        {
            lifeParameters it;
            
            it = LifeParametersFromColor(itsColor);

            // combats only occur when trophic levels are differents by one level

            // I'm a predator
            if (me.trophicLevel - it.trophicLevel == 1)
            {
                if (me.attackLevel > it.defenseLevel)
                {
                    // activate if necessary
                    if (!this.IsActivated())
                        me.energyLevel = ENERGY_LEVEL_AT_BIRTH;

                    me.energyLevel += ENERGY_INCREMENT_KILLING;                    // <---- ENERGY INCREASE, TO BE ADJUSTED
                    return KILL;
                }
            }

            // I'm a prey
            if (me.trophicLevel - it.trophicLevel == -1)
            {
                if (me.defenseLevel < it.attackLevel)
                {
                    me.energyLevel = 0;
                    return DIE;
                }
            }

            return DRAW;
                    
        }

        // Assign lifeParameters values from color
        // assume color is of a valid type
        // exception if trophicLevel is not valid
        private static lifeParameters LifeParametersFromColor(Color col)
        {
            lifeParameters pars = new lifeParameters();

            // backup
            pars.myColor = col;

            if (col.R == 255 && col.G == 255)
            {
                pars.trophicLevel = SOIL;
            }
            // black color is treated as SOIL
            else if (col.R == 0 && col.G == 0 && col.R == 0)
            {
                pars.trophicLevel = SOIL;
            }
            // a plant
            else if (col.R <= ATTACK_LEVEL_MAX && col.G >= ENERGY_LEVEL_COLOR_MIN && col.B <= DEFENSE_LEVEL_MAX)
            {
                pars.trophicLevel = PLANT;
            }
            // a herbivore
            else if (col.R <= DEFENSE_LEVEL_MAX && col.G <= ATTACK_LEVEL_MAX && col.B >= ENERGY_LEVEL_COLOR_NOT_ACTIVATED)
            {
                pars.trophicLevel = HERBIVORE;
            }
            // a carnivore
            else if (col.R >= ENERGY_LEVEL_COLOR_NOT_ACTIVATED && col.G <= DEFENSE_LEVEL_MAX && col.B <= ATTACK_LEVEL_MAX)
            {
                pars.trophicLevel = CARNIVORE;
            }
            else
            {
                // something is wrong
                throw new IndexOutOfRangeException();
            }

            switch (pars.trophicLevel)
            {
                case SOIL:
                    pars.energyLevel = 0;
                    pars.attackLevel = 0;
                    pars.defenseLevel = 128- col.B/2;
                    break;
                case PLANT:
                    pars.energyLevel = col.G - ENERGY_LEVEL_COLOR_MIN;
                    pars.attackLevel = col.R;
                    pars.defenseLevel = col.B;
                    break;
                case HERBIVORE:
                    pars.energyLevel = col.B - ENERGY_LEVEL_COLOR_MIN;
                    pars.attackLevel = col.G;
                    pars.defenseLevel = col.R;
                    break;
                case CARNIVORE:
                    pars.energyLevel = col.R - ENERGY_LEVEL_COLOR_MIN;
                    pars.attackLevel = col.B;
                    pars.defenseLevel = col.G;
                    break; 
            }

            return pars;
        }

        // Herbivores and carnivores are not "activated" until they have a first interaction with predator or prey
        // SOIL is never activated
        // PLANTS are always activated
        // This is indicated with energyLevel = -1
        public bool IsActivated()
        {
            if (me.trophicLevel == SOIL)
                return false;
            else if (me.trophicLevel == PLANT)
                return true;
            else
                return me.energyLevel > -1;
        }

        public bool IsDead()
        {
            if (me.trophicLevel == SOIL)
                return false;
            else
                return me.energyLevel == 0;
        }

        public static bool IsDead(Color c)
        {
            return (c.G == deadCreatureColor.G && c.B == deadCreatureColor.B &&  c.R == deadCreatureColor.R);               // <-- TO BE ADJUSTED
        }


        public Color GetOffspring()
        {
            int p = -15+ rand.Next(30);
            me.energyLevel = ENERGY_LEVEL_AT_BIRTH + p;                 // <--- TO BE INCLUDED: mutations
            return ColorFromLifeParameters(me);
        }

        private static Color ColorFromLifeParameters(lifeParameters me)
        {
            // for SOIL just return original color
            if (me.trophicLevel == SOIL)
                return me.myColor;

            // if creature has died, cell becames white (to avoid problems with soil defence)   <----- TO BE ADJUSTED 
            if (me.energyLevel == 0)                        // -1 means not activated
                return deadCreatureColor; 

            // format color depending on thropic level
            int e;
            if (me.energyLevel == -1)    // non activated
                e = ENERGY_LEVEL_COLOR_NOT_ACTIVATED;
            else
                e = (int)Math.Ceiling(ENERGY_LEVEL_COLOR_MIN + me.energyLevel);
            switch (me.trophicLevel)
            {
                case PLANT:
                    return Color.FromArgb(me.attackLevel, e, me.defenseLevel);
                case HERBIVORE:
                    return Color.FromArgb(me.defenseLevel, me.attackLevel, e);
                case CARNIVORE:
                    return Color.FromArgb(e, me.defenseLevel, me.attackLevel);
                default:
                    // something is wrong
                    throw new IndexOutOfRangeException();
            }
        }


        public bool IsTimeForMoving()
        {
            float p = rand.Next(1000) / 1000f;
            switch (me.trophicLevel)
            {
                case PLANT:
                    return p < PLANT_MOVE_PROBABILITY;
                case HERBIVORE:
                    return p < HERBIVORE_MOVE_PROBABILITY; 
                case CARNIVORE:
                    return p < CARNIVORE_MOVE_PROBABILITY; 

            }
            return false;
        }

        public bool IsTimeForReproduction()
        {
            float p = rand.Next(1000) / 1000f;
            switch(me.trophicLevel)
            {
                case PLANT:
                    return p < PLANT_REPRODUCTION_PROBABILITY;
                case HERBIVORE:
                    return p < HERBIVORE_REPRODUCTION_PROBABILITY; // <----  && me.energyLevel > 75;
                case CARNIVORE:
                    return p < CARNIVORE_REPRODUCTION_PROBABILITY; // <----  && me.energyLevel > 75;
 
            }
            return false;
        }

        // returns offspring color or white if it is not time for reproduction
        public static Color IsTimeForReproduction(Color c2)
        {
            Random r = new Random();
            float p = r.Next(1000) / 1000f;
            float q = 0f;
            lifeParameters pars = new lifeParameters();
            pars = LifeParametersFromColor(c2);
            pars.energyLevel = ENERGY_LEVEL_AT_BIRTH;                 // <--- TO BE INCLUDED: mutations
            switch (pars.trophicLevel)
            {
                case PLANT:
                    q = PLANT_REPRODUCTION_PROBABILITY;
                    break;
                case HERBIVORE:
                    q = HERBIVORE_REPRODUCTION_PROBABILITY;
                    break;
                case CARNIVORE:
                    q = CARNIVORE_REPRODUCTION_PROBABILITY;
                    break;

            }
            if (p < q)
            {
                return ColorFromLifeParameters(pars);
            }
            else
                return deadCreatureColor;
        }


        /// <summary>
        /// returns trophic level for the color
        /// </summary>

        public int GetTrophicLevel(Color c)
        {
            lifeParameters lp;
            lp = LifeParametersFromColor(c);
            return lp.trophicLevel;
        }
        public String GetTrophicLevelString(Color c)
        {
            lifeParameters lp;
            lp = LifeParametersFromColor(c);
            switch(lp.trophicLevel)
            {
                case SOIL:
                    return "SOIL (" + c.ToString() + ")";
                case PLANT:
                    return "PLANT (" + c.ToString() + ")";
                case HERBIVORE:
                    return "HERBIVORE (" + c.ToString() + ")";
                case CARNIVORE:
                    return "CARNIVORE (" + c.ToString() + ")";
                default:
                    return "UNKNOW ("+c.ToString() +")";
            }
        }

        /// <summary>
        /// returns creature's color from parameters
        /// </summary>

        public static Color GetColor(int trophicLevel, int energyLevel, int attackLevel, int defenseLevel)
        {
            lifeParameters me = new lifeParameters();
            me.trophicLevel = trophicLevel;
            me.energyLevel = energyLevel;
            me.attackLevel = attackLevel;
            me.defenseLevel = defenseLevel;
            me.myColor = ColorFromLifeParameters(me);
            return me.myColor;
        }


        // Changes color to reflect energy level at the beginning of simulation
        // Plants have ENERGY_LEVEL_AT_BIRTH
        // Herbivores and Carnivores have -1. They don't start losing energy until they eat something
        // Return White if color is not a valid creature
        public static Color SetInitialEnergyLevel(Color c)
        {
            try
            {
                lifeParameters myPars = LifeParametersFromColor(c);
                switch (myPars.trophicLevel)
                {
                    case SOIL:
                        break;
                    case PLANT:
                        myPars.energyLevel = ENERGY_LEVEL_AT_BIRTH;
                        break;
                    case HERBIVORE:
                        myPars.energyLevel = -1;            // non activated
                        break;
                    case CARNIVORE:
                        myPars.energyLevel = -1;            // non activated
                        break;
                }

                return ColorFromLifeParameters(myPars);
            }
            catch (Exception)
            {
                return deadCreatureColor;
            }
        }

 
    }
}

File: src/Creature.ts

export type Color = {
    r: number;
    g: number;
    b: number;
}

export function colorEqual(c1: Color, c2: Color): boolean {
    return c1.r==c2.r && c1.g == c2.g && c1.b == c2.b;
}


//  DEFINES 

export enum ThrophicLevel {    
    SOIL, 
    PLANT,
    HERBIVORE,
    CARNIVORE
}

export enum AttackResult {
    DRAW,
    KILL,
    DIE
}
  // LifeParameters can be represented by a RGB color
export type LifeParameters = {
    trophicLevel : number ,
    attackLevel : number ,
    defenseLevel : number ,
    energyLevel : number ,
    dead : boolean,
    activated : boolean,      // non activated creatures doesn't have metabolism. soil is never  activated.
                            // it prevents animals die before they interact with some prey or predators
    color : Color
}

export class Creature {

    static deadCreatureColor: Color = {r: 255, g: 255, b: 255};    // equivalent to less defense soil
    static soilColor: Color = {r: 255, g: 255, b: 128};   
    static errorColor: Color = {r: 240, g: 50, b: 12};   
    static blackColor: Color = {r: 0, g: 0, b: 0};   
    

    //
    //  SIMULATION PARAMETERS
    //

    // consumes this energy every cycle
    static ENERGY_CONSUMPTION_PLANTS: number = 1;
    static ENERGY_CONSUMPTION_HERBIVORES: number = 2;       // 0.5
    static ENERGY_CONSUMPTION_CARNIVORES: number = 2;

    static ENERGY_INCREMENT_KILLING: number = 2;

    static DEFENSE_LEVEL_MAX: number = 150;
    static ATTACK_LEVEL_MAX: number = 150;
    static ENERGY_LEVEL_MAX: number = 100;
    static ENERGY_LEVEL_COLOR_MIN: number = 155;              // energyLevel = color - ENERGY_LEVEL_COLOR_MIN
    static ENERGY_LEVEL_COLOR_NOT_ACTIVATED: number = 154;   // 154 (energyLevel = -1) means "not activated"


    static ENERGY_LEVEL_AT_BIRTH: number = 50;

    static PLANT_REPRODUCTION_PROBABILITY: number = 0.2;        // recommended 0.1 - 0.5
    static HERBIVORE_REPRODUCTION_PROBABILITY: number = 0.2;    // recommended 0.15 - 
    static CARNIVORE_REPRODUCTION_PROBABILITY: number = 0.1;

    // probability to move after killing a creature
    static PLANT_MOVE_PROBABILITY: number = 0;        
    static HERBIVORE_MOVE_PROBABILITY: number = 0.2;     
    static CARNIVORE_MOVE_PROBABILITY: number = 0.4;



    // initialize a creature with this color and do rutine actions
    // Consumes energy, updates color 
    // If herbivores or carnivores are not still activated, don't consume energy
    static beginIteration(lp_in: LifeParameters): LifeParameters {
        let lp : LifeParameters = lp_in;
        
        // if activated consume energy 
        if (lp.activated && !lp.dead) {
            switch (lp.trophicLevel)
            {
                case ThrophicLevel.PLANT:
                    lp.energyLevel -= Creature.ENERGY_CONSUMPTION_PLANTS;
                    break;
                case ThrophicLevel.HERBIVORE:
                    lp.energyLevel -= Creature.ENERGY_CONSUMPTION_HERBIVORES;
                    break;
                case ThrophicLevel.CARNIVORE:
                    lp.energyLevel -= Creature.ENERGY_CONSUMPTION_CARNIVORES;
                    break;
            }
            lp.energyLevel = lp.energyLevel < 0 ? 0 : lp.energyLevel;
        }

        // set my color with new energy level
        lp.color = this.colorFromLifeParameters(lp);
        
        return lp;
    }

    
    // returns attackResult

    static attack(me: LifeParameters, it: LifeParameters) : { result: AttackResult; me_post: LifeParameters; it_post: LifeParameters;}  {

        let res : AttackResult = AttackResult.DRAW;

            // <--- QUE PASSA SI UN ESTA MORT?

            // combats only occur when trophic levels are differents by one level

            // I'm a predator
            if (me.trophicLevel == ThrophicLevel.CARNIVORE && it.trophicLevel == ThrophicLevel.HERBIVORE 
                || me.trophicLevel == ThrophicLevel.HERBIVORE && it.trophicLevel == ThrophicLevel.PLANT
                || me.trophicLevel == ThrophicLevel.PLANT && it.trophicLevel == ThrophicLevel.SOIL) {

                if (me.attackLevel > it.defenseLevel)
                {
                    // activate if necessary
                    if (!me.activated) {
                            me.activated = true;
                            me.energyLevel = Creature.ENERGY_LEVEL_AT_BIRTH;    // <-- HAURIA DE SER DIFERENT SEGONS TL?
                        }
                    me.energyLevel += Creature.ENERGY_INCREMENT_KILLING;    
                    it.energyLevel = 0;     
                    res = AttackResult.KILL;
                }
            }
            // I'm a prey
            else if (it.trophicLevel == ThrophicLevel.CARNIVORE && me.trophicLevel == ThrophicLevel.HERBIVORE 
                || it.trophicLevel == ThrophicLevel.HERBIVORE && me.trophicLevel == ThrophicLevel.PLANT
                || it.trophicLevel == ThrophicLevel.PLANT && me.trophicLevel == ThrophicLevel.SOIL) {
                if (me.defenseLevel < it.attackLevel)
                {
                    // activate if necessary
                    if (!it.activated) {
                        it.activated = true;
                        it.energyLevel = Creature.ENERGY_LEVEL_AT_BIRTH;    // <-- HAURIA DE SER DIFERENT SEGONS TL?
                    }
                    it.energyLevel += Creature.ENERGY_INCREMENT_KILLING;          
                    me.energyLevel = 0;     
                    res = AttackResult.DIE;
                }
            }

            // adjust color if energy changed
            if (res!=AttackResult.DRAW) {
                me.color = this.colorFromLifeParameters(me);      
                it.color = this.colorFromLifeParameters(it);      
            }
                    
            return  {
                result : res,
                me_post : me,
                it_post : it
            }
    }

 
    
    static lifeParametersFromColor(col: Color): LifeParameters {
        let pars : LifeParameters = {
            trophicLevel : 0,
            attackLevel : 0,
            defenseLevel : 0,
            energyLevel : 0,
            dead : false,
            activated : false,
            color : col      
        };    // <-- com inicialitzar en blanc?????

        // backup

        if (col.r == 255 && col.g == 255)
        {
            pars.trophicLevel = ThrophicLevel.SOIL;
        }
        // black color is treated as SOIL
        else if (col.r == 0 && col.g == 0 && col.b == 0)
        {
            pars.trophicLevel = ThrophicLevel.SOIL;
        }
        // a plant
        else if (col.r <= Creature.ATTACK_LEVEL_MAX && col.g >= Creature.ENERGY_LEVEL_COLOR_MIN && col.b <= Creature.DEFENSE_LEVEL_MAX)
        {
            pars.trophicLevel = ThrophicLevel.PLANT;
        }
        // a herbivore
        else if (col.r <= Creature.DEFENSE_LEVEL_MAX && col.g <= Creature.ATTACK_LEVEL_MAX && col.b >= Creature.ENERGY_LEVEL_COLOR_NOT_ACTIVATED)
        {
            pars.trophicLevel = ThrophicLevel.HERBIVORE;
        }
        // a carnivore
        else if (col.r >= Creature.ENERGY_LEVEL_COLOR_NOT_ACTIVATED && col.g <= Creature.DEFENSE_LEVEL_MAX && col.b <= Creature.ATTACK_LEVEL_MAX)
        {
            pars.trophicLevel = ThrophicLevel.CARNIVORE;
        }
        else
        {
            // something is wrong
            console.log("lifeParametersFromColor error color:", col, "pars:", pars);
        }

        switch (pars.trophicLevel)
        {
            case ThrophicLevel.SOIL:
                pars.energyLevel = 0;
                pars.attackLevel = 0;
                pars.defenseLevel = 128- col.b/2;
                pars.activated = false;
                break;
            case ThrophicLevel.PLANT:
                pars.energyLevel = col.g - Creature.ENERGY_LEVEL_COLOR_MIN;
                pars.attackLevel = col.r;
                pars.defenseLevel = col.b;
                pars.activated = true;
                break;
            case ThrophicLevel.HERBIVORE:
                pars.energyLevel = col.b - Creature.ENERGY_LEVEL_COLOR_MIN;
                pars.attackLevel = col.g;
                pars.defenseLevel = col.r;
                pars.activated = pars.energyLevel >  -1;
                break;
            case ThrophicLevel.CARNIVORE:
                pars.energyLevel = col.r - Creature.ENERGY_LEVEL_COLOR_MIN;
                pars.attackLevel = col.b;
                pars.defenseLevel = col.g;
                pars.activated = pars.energyLevel >  -1;
                break; 
        }

        pars.dead = pars.energyLevel == 0;

        return pars;
    }

 

    static getOffspring(lp:LifeParameters): LifeParameters {
        let p : number = -15+ 30*Math.random();
        lp.energyLevel = Creature.ENERGY_LEVEL_AT_BIRTH + p;                 // <--- TO BE INCLUDED: mutations
        lp.activated = true;           // offspring are always activated    <--- aqui no fa res aixo
        lp.color = Creature.colorFromLifeParameters(lp);
        return lp;
    }



    static isTimeForMoving(me:LifeParameters) : boolean {
        let p : number = Math.random();
        switch (me.trophicLevel)
        {
            case ThrophicLevel.PLANT:
                return p < Creature.PLANT_MOVE_PROBABILITY;
            case ThrophicLevel.HERBIVORE:
                return p < Creature.HERBIVORE_MOVE_PROBABILITY; 
            case ThrophicLevel.CARNIVORE:
                return p < Creature.CARNIVORE_MOVE_PROBABILITY; 

        }
        return false;
    }

    
    static isTimeForReproduction(me:LifeParameters): boolean{
        let p : number = Math.random();
        switch(me.trophicLevel)
        {
            case ThrophicLevel.PLANT:
                return p < Creature.PLANT_REPRODUCTION_PROBABILITY;
            case ThrophicLevel.HERBIVORE:
                return p < Creature.HERBIVORE_REPRODUCTION_PROBABILITY; // <----  && me.energyLevel > 75;
            case ThrophicLevel.CARNIVORE:
                return p < Creature.CARNIVORE_REPRODUCTION_PROBABILITY; // <----  && me.energyLevel > 75;

        }
        return false;
    }
    
   

    static isSoil(c: Color): boolean {
        return c.r == 255 && c.g == 255;
    }


    // returns creature's color from parameters
    static getColor(trophicLevel: ThrophicLevel, energyLevel: number, attackLevel: number, defenseLevel: number): Color {
        let me : LifeParameters = {
        trophicLevel : trophicLevel,
        energyLevel : energyLevel,
        attackLevel : attackLevel,
        defenseLevel : defenseLevel,
        dead : energyLevel == 0,
        activated : trophicLevel == ThrophicLevel.PLANT || (trophicLevel != ThrophicLevel.SOIL && energyLevel>-1),
        color : Creature.errorColor
        }
        return this.colorFromLifeParameters(me);
    }

    static colorFromLifeParameters(me: LifeParameters) : Color {
        if (me.trophicLevel == ThrophicLevel.SOIL)
            return Creature.soilColor;     // <--- ojo tenia energia diferent ?

        // if creature has died, cell becames white
        if (me.energyLevel == 0)                       
            return Creature.deadCreatureColor; 

        // format color depending on thropic level
        let e : number;
        if (me.energyLevel == -1)    // non activated
            e = Creature.ENERGY_LEVEL_COLOR_NOT_ACTIVATED;
        else
            e = Math.ceil(Creature.ENERGY_LEVEL_COLOR_MIN + me.energyLevel);
        let c : Color;
        // code into RGB depending on trophiclevel
        switch (me.trophicLevel)
        {
            case ThrophicLevel.PLANT:
                c = {r: me.attackLevel, g: e, b: me.defenseLevel};
                break;
            case ThrophicLevel.HERBIVORE:
                c = {r: me.defenseLevel, g: me.attackLevel, b: e};
                break;
            case ThrophicLevel.CARNIVORE:
                c = {r: e, g: me.defenseLevel, b: me.attackLevel};
                break;
            default:
                // something is wrong
                console.log("Creature > colorFromLifeParameters  *** Error ***"+me.toString())
                c = Creature.errorColor;  
        }            
        return c;

    } 


}